pageserver/src/tenant.rs

```diff
@@ -3258,6 +3260,58 @@ impl Tenant {
     pub fn cached_synthetic_size(&self) -> u64 {
         self.cached_synthetic_tenant_size.load(Ordering::Relaxed)
     }
+
+    /// Flush any in-progress layers, schedule uploads, and wait for uploads to complete.
+    ///
+    /// This function can take a long time: callers should wrap it in a timeout if calling
+    /// from an external API handler.
+    ///
+    /// Cancel-safety: cancelling this function may leave I/O running, but such I/O is
+    /// still bounded by tenant/timeline shutdown.
+    #[tracing::instrument(skip_all)]
+    pub(crate) async fn flush_remote(&self) -> anyhow::Result<()> {
+        let timelines = self.timelines.lock().unwrap().clone();
+
+        async fn flush_timeline(_gate: GateGuard, timeline: Arc<Timeline>) -> anyhow::Result<()> {
+            tracing::info!(timeline_id=%timeline.timeline_id, "Flushing...");
+            timeline.freeze_and_flush().await?;
+            tracing::info!(timeline_id=%timeline.timeline_id, "Waiting for uploads...");
+            if let Some(client) = &timeline.remote_client {
+                client.wait_completion().await?;
+            }
+
+            Ok(())
+        }
+
+        let mut tasks = tokio::task::JoinSet::new();
+        for (_timeline_id, timeline) in timelines {
+            // Run each timeline's flush in a task holding the timeline's gate: this
+            // means that if this function's future is cancelled, the Timeline shutdown
+            // will still wait for any I/O in here to complete, enforcing the safety
+            // invariant that after Timeline::shutdown nothing touches the timeline's
+            // local storage.
+            let gate = match timeline.gate.enter() {
+                Ok(g) => g,
+                Err(_) => continue,
+            };
+            tasks.spawn(async move { flush_timeline(gate, timeline).await });
+        }
+
+        while let Some(r) = tasks.join_next().await {
+            let _ = r?;
```

Comment:
```suggestion
            if let Err(e) = r {
                if !e.is_cancel() && !e.is_panic() {
                    tracing::error!("unexpected joinset error: {e:?}");
                }
            }
```

Reply:
Why am I suggesting not to use a `?`, because there's:
- JoinError::is_cancelled -- cannot really happen because we are not using abort handle (which is returned from JoinSet::spawn)
- JoinError::is_panic -- it has been logged already, but would it be a good enough to return from here? the change has already been made